internal/ls/sourcemaphost.go

```diff
@@ -0,0 +1,197 @@
+	host := &sourcemapHost{program: program}
+	fileReader := &sourcemapFileReader{host: host}
+	return sourcemap.CreateSourceMapper(host, fileReader)
+}
+
+
+// Maps a single definition location using source maps.
+func MapSingleDefinitionLocation(program *compiler.Program, location lsproto.Location, languageService *LanguageService) *lsproto.Location {
+	fileName := location.Uri.FileName()
+
+	if !strings.HasSuffix(fileName, ".d.ts") {
+		return nil
+	}
+
+	host := &sourcemapHost{program: program}
+	sourceMapper := CreateSourceMapperForProgram(program)
+	
+	return tryMapLocation(sourceMapper, host, location, languageService)
+}
+
+func tryMapLocation(sourceMapper sourcemap.SourceMapper, host *sourcemapHost, location lsproto.Location, languageService *LanguageService) *lsproto.Location {
+	fileName := location.Uri.FileName()
+	program := host.program
+	
+	declFile := program.GetSourceFile(fileName)
+	var declStartPos, declEndPos int
+	
+	if declFile != nil {
+		declStartPos = int(languageService.converters.LineAndCharacterToPosition(declFile, location.Range.Start))
+		declEndPos = int(languageService.converters.LineAndCharacterToPosition(declFile, location.Range.End))
+	} else {
+		declContent, ok := host.readFileWithFallback(fileName)
+		if !ok {
+			return nil
+		}
+		
+		declScript := &textScript{
+			fileName: fileName,
+			text:     declContent,
+		}
+		
+		declStartPos = int(languageService.converters.LineAndCharacterToPosition(declScript, location.Range.Start))
+		declEndPos = int(languageService.converters.LineAndCharacterToPosition(declScript, location.Range.End))
+	}
+	
+	startInput := sourcemap.DocumentPosition{
+		FileName: fileName,
+		Pos:      declStartPos,
+	}
+
+	startResult := sourceMapper.TryGetSourcePosition(startInput)
+	if startResult == nil {
+		return nil
+	}
+
+	originalRangeLength := declEndPos - declStartPos
+	sourceEndPos := startResult.Pos + originalRangeLength
+	
+	sourceFile := program.GetSourceFile(startResult.FileName)
+	var sourceStartLSP lsproto.Position
+	var sourceEndLSP lsproto.Position
+	
+	if sourceFile != nil {
+		if sourceEndPos > len(sourceFile.Text()) {
+			sourceEndPos = len(sourceFile.Text())
+		}
+		
+		sourceStartLSP = languageService.converters.PositionToLineAndCharacter(sourceFile, core.TextPos(startResult.Pos))
+		sourceEndLSP = languageService.converters.PositionToLineAndCharacter(sourceFile, core.TextPos(sourceEndPos))
+
+	} else {
+		sourceContent, ok := host.readFileWithFallback(startResult.FileName)
+		if !ok {
+			return nil
+		}
+		
+		if sourceEndPos > len(sourceContent) {
+			sourceEndPos = len(sourceContent)
+		}
+		
+		textScript := &textScript{
+			fileName: startResult.FileName,
+			text:     sourceContent,
+		}
+		
+		sourceStartLSP = languageService.converters.PositionToLineAndCharacter(textScript, core.TextPos(startResult.Pos))
+		sourceEndLSP = languageService.converters.PositionToLineAndCharacter(textScript, core.TextPos(sourceEndPos))
```

Comment:
These two blocks (if and else are the same); I would instead declare `textScript` outside, and then conditionally assign it, then move these two converter lines into the struct below.